### PR TITLE
Pv2 1337 update the d lock actor to be scoped to uln rather than ukprn

### PR DESCRIPTION
**Pull request by abroodie**

### CODE REVIEW

File: src/SFA.DAS.Payments.PerformanceTests/TestsBase.cs [hunk 1]
```diff
@@ -42,7 +43,7 @@ protected FM36Learner CreateFM36Learner(TestSession session, Learner testLearner
                     PriceEpisodeFundLineType = testLearner.Course.FundingLineType,
                     PriceEpisodeBalanceValue = 0,
                     PriceEpisodeCompletionPayment = 3000,
-                    PriceEpisodeContractType = "Non-Levy Contract", //ContractType.Act2.ToString("G"),
+                    PriceEpisodeContractType = testLearner.IsLevyLearner ? "Levy Contract" : "Non-Levy Contract", //ContractType.Act2.ToString("G"),
```
Comment (davehowlett) [inline src/SFA.DAS.Payments.PerformanceTests/TestsBase.cs]: remove the commented out code?

File: src/SFA.DAS.Payments.RequiredPayments.RequiredPaymentsService/RequiredPaymentsService.cs [hunk 2]
```diff
@@ -41,7 +43,7 @@ public class RequiredPaymentsService : Actor, IRequiredPaymentsService
             ActorId actorId,
             IPaymentLogger paymentLogger,
             IApprenticeshipKeyService apprenticeshipKeyService,
-            IPaymentHistoryRepository paymentHistoryRepository,
+            Func<IPaymentHistoryRepository> paymentHistoryRepositoryFactory,
             IPaymentKeyService paymentKeyService,
```
Comment (davehowlett) [inline src/SFA.DAS.Payments.RequiredPayments.RequiredPaymentsService/RequiredPaymentsService.cs]: this isn't used. Can it be removed?

File: src/SFA.DAS.Payments.DataLocks.DataLockProxyService/Handlers/ApprenticeshipContractType1EarningEventHandler.cs [hunk 3]
```diff
@@ -15,38 +15,52 @@ namespace SFA.DAS.Payments.DataLocks.DataLockProxyService.Handlers
...
-            paymentLogger.LogInfo($"Processing DataLockProxyProxyService event for UKPRN {message.Ukprn}");
+            try
+            {
+                if (message.Learner== null || message.Learner?.Uln == 0)
+                {
+                    throw new InvalidOperationException("Invalid 'ApprenticeshipContractType1EarningEvent' received. Learner was null or Uln was 0.");
+                }
+                var uln = message.Learner.Uln;
+                var learnerRef = message.Learner.ReferenceNumber;
+                logger.LogDebug($"Processing DataLockProxyProxyService event for learner with learn ref {learnerRef}");
```
[the remark this answers is not in the document]
Comment (abroodie) [reply src/SFA.DAS.Payments.DataLocks.DataLockProxyService/Handlers/ApprenticeshipContractType1EarningEventHandler.cs]: it doesn't state that the learner ref is the actor id, just says it creatnig an actor for that learner.  Updated the text in the logging.

File: src/SFA.DAS.Payments.DataLocks.DataLockProxyService/Handlers/ApprenticeshipContractType1EarningEventHandler.cs [hunk 4]
```diff
@@ -15,38 +15,52 @@ namespace SFA.DAS.Payments.DataLocks.DataLockProxyService.Handlers
...
-            var ukprn = message.Ukprn;
-            var actorId = new ActorId(ukprn);
+                logger.LogVerbose($"Creating actor proxy for learner with leearner ref {learnerRef}");
+                var actor = proxyFactory.CreateActorProxy<IDataLockService>(new Uri("fabric:/SFA.DAS.Payments.DataLocks.ServiceFabric/DataLockServiceActorService"), actorId);
+                logger.LogDebug($"Actor proxy created for learner with ULN {uln}");
 
-            paymentLogger.LogVerbose($"Creating actor proxy for provider with UKPRN {ukprn}");
-            var actor = proxyFactory.CreateActorProxy<IDataLockService>(new Uri("fabric:/SFA.DAS.Payments.DataLocks.ServiceFabric/DataLockServiceActorService"), actorId);
-            paymentLogger.LogDebug($"Actor proxy created for UKPRN {ukprn}");
+                logger.LogVerbose($"Calling actor proxy to handle earning for learner with leearner ref {learnerRef}");
```
[the remark this answers is not in the document]
Comment (abroodie) [reply src/SFA.DAS.Payments.DataLocks.DataLockProxyService/Handlers/ApprenticeshipContractType1EarningEventHandler.cs]: fixed

File: src/SFA.DAS.Payments.DataLocks.DataLockProxyService/Handlers/ApprenticeshipUpdatedHandler.cs [hunk 5]
```diff
@@ -25,17 +25,22 @@ public async Task Handle(ApprenticeshipUpdated message, IMessageHandlerContext c
         {
             try
             {
-                logger.LogDebug($"Now handling the apprenticeship updated event.  Apprenticeship: {message.Id}, employer: {message.EmployerAccountId}, provider: {message.Ukprn}");
+                if (message.Uln == 0)
+                {
+                    throw new InvalidOperationException("Invalid 'ApprenticeshipUpdated' received. Uln was 0.");
+                }
+                
+                logger.LogDebug($"Now handling the apprenticeship updated event.  Apprenticeship: {message.Id}, employer: {message.EmployerAccountId}, ukprn: {message.Ukprn}, learner with ULN {message.Uln.ToString().Substring(0, 4)}");
```
[the remark this answers is not in the document]
Comment (abroodie) [reply src/SFA.DAS.Payments.DataLocks.DataLockProxyService/Handlers/ApprenticeshipUpdatedHandler.cs]: fixed

File: src/SFA.DAS.Payments.DataLocks.DataLockProxyService/Handlers/ResetCacheHandler.cs [hunk 6]
```diff
@@ -1,35 +1,76 @@
...
 using Microsoft.ServiceFabric.Actors.Client;
 using NServiceBus;
 using SFA.DAS.Payments.Application.Infrastructure.Logging;
 using SFA.DAS.Payments.DataLocks.DataLockService.Interfaces;
+using SFA.DAS.Payments.DataLocks.Domain.Services.Apprenticeships;
 using SFA.DAS.Payments.DataLocks.Messages.Internal;
 
 namespace SFA.DAS.Payments.DataLocks.DataLockProxyService.Handlers
 {
+    public class ResetActorsHandler: IHandleMessages<ResetActorsCommand>
```
[the remark this answers is not in the document]
Comment (abroodie) [reply src/SFA.DAS.Payments.DataLocks.DataLockProxyService/Handlers/ResetCacheHandler.cs]: done

File: src/SFA.DAS.Payments.DataLocks.DataLockProxyService/Handlers/ResetCacheHandler.cs [hunk 7]
```diff
@@ -1,35 +1,76 @@
...
             {
                 logger.LogDebug($"Resetting cache for provider :{message.Ukprn}");
-                var actorId = new ActorId(message.Ukprn);
-                logger.LogVerbose($"Creating actor proxy for provider with UKPRN {message.Ukprn}");
-                var actor = proxyFactory.CreateActorProxy<IDataLockService>(new Uri("fabric:/SFA.DAS.Payments.DataLocks.ServiceFabric/DataLockServiceActorService"), actorId);
-                logger.LogDebug($"Actor proxy created for UKPRN {message.Ukprn}, now resetting the cache.");
-                await actor.Reset().ConfigureAwait(false);
+                var ulns = await repository.ApprenticeshipUlnsByProvider(message.Ukprn);
+                var resetTasks = new List<Task>();
+                foreach (var uln in ulns)
```
[the remark this answers is not in the document]
Comment (abroodie) [reply src/SFA.DAS.Payments.DataLocks.DataLockProxyService/Handlers/ResetCacheHandler.cs]: ResetCache will be removed in a subsequent story

File: src/SFA.DAS.Payments.Monitoring.Jobs.Client/JobMessageClient.cs [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+        {
+            this.messageSession = messageSession ?? throw new ArgumentNullException(nameof(messageSession));
+            this.logger = logger ?? throw new ArgumentNullException(nameof(logger));
+        }
+
+
+        public async Task ProcessedJobMessage(long jobId, Guid messageId, string messageName, List<GeneratedMessage> generatedMessages)
+        {
+            logger.LogVerbose($"Sending request to record successful processing of event. Job Id: {jobId}, Event: id: {messageId} ");
+            var itemProcessedEvent = new RecordJobMessageProcessingStatus
```
[the remark this answers is not in the document]
Comment (abroodie) [reply src/SFA.DAS.Payments.Monitoring.Jobs.Client/JobMessageClient.cs]: Monitoring service handles this message

File: src/SFA.DAS.Payments.Monitoring.Jobs.Client/JobMessageClientFactory.cs [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+using Autofac;
+
+namespace SFA.DAS.Payments.Monitoring.Jobs.Client
+{
+    public interface IJobMessageClientFactory
+    {
+        IJobMessageClient Create();
+    }
+
+    public class JobMessageClientFactory : IJobMessageClientFactory
```
[the remark this answers is not in the document]
Comment (abroodie) [reply src/SFA.DAS.Payments.Monitoring.Jobs.Client/JobMessageClientFactory.cs]: I'd rather keep them separate to make it easier to test.  open to discuss in a dev design session though.  Create a tech ticket for it.

File: src/SFA.DAS.Payments.PerformanceTests/TestsBase.cs [hunk 10]
```diff
@@ -20,10 +20,10 @@ namespace SFA.DAS.Payments.PerformanceTests
     public class TestsBase
     {
         public static TestsConfiguration Config => new TestsConfiguration();
-        protected FM36Learner CreateFM36Learner(TestSession session, Learner testLearner)
+        protected FM36Learner CreateFM36Learner(TestSession session, Learner testLearner, DateTime startDate)
         {
             var learner = new FM36Learner { LearnRefNumber = testLearner.LearnRefNumber, ULN = testLearner.Uln};
-            var startDate = new DateTime(DateTime.Today.Year + (DateTime.Today.Month < 8 ? -1 : 0), 8, 1);
+            //var startDate = new DateTime(DateTime.Today.Year + (DateTime.Today.Month < 8 ? -1 : 0), 8, 1);
```
[the remark this answers is not in the document]
Comment (abroodie) [reply src/SFA.DAS.Payments.PerformanceTests/TestsBase.cs]: done

File: src/SFA.DAS.Payments.PerformanceTests/TestsBase.cs (hunk 1 above)
Comment (abroodie) [reply src/SFA.DAS.Payments.PerformanceTests/TestsBase.cs]: done

File: src/SFA.DAS.Payments.RequiredPayments.RequiredPaymentsService/RequiredPaymentsService.cs (hunk 2 above)
Comment (abroodie) [reply src/SFA.DAS.Payments.RequiredPayments.RequiredPaymentsService/RequiredPaymentsService.cs]: Done

File: src/SFA.DAS.Payments.Monitoring.Jobs.Client/JobMessageClient.cs [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+            {
+                JobId = jobId,
+                Id = messageId,
+                MessageName = messageName,
+                EndTime = DateTimeOffset.UtcNow,
+                GeneratedMessages = generatedMessages ?? new List<GeneratedMessage>(),
+                Succeeded = true
```
[the remark this answers is not in the document]
Comment (abroodie) [reply src/SFA.DAS.Payments.Monitoring.Jobs.Client/JobMessageClient.cs]: Will be used when the monitoring client hooks into NSB notifications to get details of messages sent to the error queues.

File: src/SFA.DAS.Payments.Monitoring.Jobs.Client/Infrastructure/Ioc/JobStatusClientModule.cs [hunk 12]
```diff
@@ -49,5 +78,36 @@ private void Errors_MessageSentToErrorQueue(object sender, NServiceBus.Faults.Fa
...
+                .ConnectionString(configHelper.GetConnectionString("MonitoringServiceBusConnectionString"))
+                .Transactions(TransportTransactionMode.ReceiveOnly)
+                .RuleNameShortener(ruleName => ruleName.Split('.').LastOrDefault() ?? ruleName);
+
+            transport.Routing().RouteToEndpoint(typeof(RecordStartedProcessingEarningsJob).Assembly, jobsEndpointName);
+            endpointConfiguration.SendFailedMessagesTo(config.FailedMessagesQueue);
+            endpointConfiguration.UseSerialization<NewtonsoftSerializer>();
+            endpointConfiguration.EnableInstallers();
+
+            endpointConfiguration.RegisterComponents(cfg => cfg.RegisterSingleton((IPaymentLogger)logger));
```
[the remark this answers is not in the document]
Comment (abroodie) [reply src/SFA.DAS.Payments.Monitoring.Jobs.Client/Infrastructure/Ioc/JobStatusClientModule.cs]: done

File: src/SFA.DAS.Payments.Monitoring.Jobs.Client/Infrastructure/Ioc/JobStatusClientModule.cs [hunk 13]
```diff
@@ -1,40 +1,69 @@
...
+                    var logger = c.Resolve<IPaymentLogger>();
+                    var factory = c.Resolve<IMonitoringMessageSessionFactory>();
+                    return new JobMessageClient(factory.Create(), logger);
+                })
+                .As<IJobMessageClient>()
                 .SingleInstance();
 
-            builder.RegisterType<JobStatusOutgoingMessageBehaviour>()
+            builder.RegisterType<JobMessageClientFactory>()
+                .As<IJobMessageClientFactory>()
```
[the remark this answers is not in the document]
Comment (abroodie) [reply src/SFA.DAS.Payments.Monitoring.Jobs.Client/Infrastructure/Ioc/JobStatusClientModule.cs]: no, doesn't need to be.

File: src/SFA.DAS.Payments.Monitoring.Jobs.Client/Infrastructure/Ioc/JobStatusClientModule.cs [hunk 14]
```diff
@@ -1,40 +1,69 @@
...
+                    var logger = c.Resolve<IPaymentLogger>();
+                    var factory = c.Resolve<IMonitoringMessageSessionFactory>();
+                    var dataContext = c.Resolve<IJobsDataContext>();
+                    return new EarningsJobClient(logger, dataContext, c.Resolve<Application.Infrastructure.Telemetry.ITelemetry>());
+                })
+                .As<IEarningsJobClient>()
+                .InstancePerDependency();
+
             builder.RegisterType<EarningsJobClientFactory>()
                 .As<IEarningsJobClientFactory>()
```
[the remark this answers is not in the document]
Comment (abroodie) [reply src/SFA.DAS.Payments.Monitoring.Jobs.Client/Infrastructure/Ioc/JobStatusClientModule.cs]: Doesn't need to be.

File: src/SFA.DAS.Payments.Monitoring.Jobs.Client/EarningsJobClient.cs [hunk 15]
```diff
@@ -1,73 +1,94 @@
...
+                await dataContext.SaveNewJob(jobDetails, jobSteps);
+                scope.Complete();
             }
-            logger.LogDebug($"Sent request to record start of earnings job. Job Id: {jobId}, Ukprn: {ukprn}");
+            SendTelemetry(jobId, ukprn, ilrSubmissionTime, collectionYear, collectionPeriod, generatedMessages.Count, jobDetails, stopwatch);
+            logger.LogInfo($"Finished saving the job to the db.  Job id: {jobDetails.Id}, DC Job Id: {jobId}, Ukprn: {ukprn}.");
         }
 
-        public async Task ProcessedJobMessage(long jobId, Guid messageId, string messageName, List<GeneratedMessage> generatedMessages)
+        private void SendTelemetry(long jobId, long ukprn, DateTime ilrSubmissionTime, short collectionYear, byte collectionPeriod, int learnerCount, JobModel jobDetails, Stopwatch stopwatch)
```
[the remark this answers is not in the document]
Comment (abroodie) [reply src/SFA.DAS.Payments.Monitoring.Jobs.Client/EarningsJobClient.cs]: Done

File: src/SFA.DAS.Payments.FundingSource.LevyFundedService/LevyFundedService.cs [hunk 16]
```diff
@@ -106,11 +108,13 @@ public async Task<ReadOnlyCollection<FundingSourcePaymentEvent>> UnableToFundTra
         {
             try
             {
-                using (var operation = telemetry.StartOperation())
+                using (var operation = telemetry.StartOperation("LevyFundedService.UnableToFundTransfer", message.EventId.ToString()))
```
[the remark this answers is not in the document]
Comment (abroodie) [reply src/SFA.DAS.Payments.FundingSource.LevyFundedService/LevyFundedService.cs]: would be better as an ServiceFabric MessageDispatcher.

File: src/SFA.DAS.Payments.EarningEvents.Application/Infrastructure/Ioc/EarningEventsModule.cs [hunk 17]
```diff
@@ -46,6 +47,10 @@ protected override void Load(ContainerBuilder builder)
             builder.RegisterType<BulkWriter<SubmittedLearnerAimModel>>()
                 .AsImplementedInterfaces()
                 .InstancePerLifetimeScope();
+
+            builder.RegisterType<SubmittedLearnerAimRepository>()
+                .AsImplementedInterfaces()
+                .InstancePerLifetimeScope();
             //builder.Register(c => new ApprenticeshipContractType2EarningEventsService
```
[the remark this answers is not in the document]
Comment (abroodie) [reply src/SFA.DAS.Payments.EarningEvents.Application/Infrastructure/Ioc/EarningEventsModule.cs]: done

File: src/SFA.DAS.Payments.Datalocks.Messages/Events/EmployerChangedProviderPriority.cs [hunk 18]
```diff
@@ -1,12 +1,21 @@
...
+    public class EmployerChangedProviderPriority : IEvent
     {
         public long EmployerAccountId { get; set; }
         public List<long> OrderedProviders { get; set; }
+        public Guid EventId { get; set; }
+        public DateTimeOffset EventTime { get; set; }
+
+        public EmployerChangedProviderPriority()
+        {
+            EventId = Guid.NewGuid();
```
[the remark this answers is not in the document]
Comment (abroodie) [reply src/SFA.DAS.Payments.Datalocks.Messages/Events/EmployerChangedProviderPriority.cs]: needed for serialisation and allow easier testing. Follows same pattern as existing events.

File: src/SFA.DAS.Payments.Datalocks.Messages/Events/ApprenticeshipUpdated.cs [hunk 19]
```diff
@@ -24,11 +26,15 @@ public class ApprenticeshipUpdated
         public ApprenticeshipStatus Status { get; set; }
         public List<ApprenticeshipPriceEpisodeModel> ApprenticeshipPriceEpisodes { get; set; }
         public List<ApprenticeshipDuplicate> Duplicates { get; set; }
+        public Guid EventId { get; set; }
+        public DateTimeOffset EventTime { get; set; }
 
         public ApprenticeshipUpdated()
         {
+            EventId = Guid.NewGuid();
```
[the remark this answers is not in the document]
Comment (abroodie) [reply src/SFA.DAS.Payments.Datalocks.Messages/Events/ApprenticeshipUpdated.cs]: needed for serialisation and allow easier testing. Follows same pattern as existing events.

File: src/SFA.DAS.Payments.DataLocks.Domain.UnitTests/Services/CourseValidation/EarningPeriodsValidationProcessorTests.cs [hunk 20]
```diff
@@ -21,6 +21,7 @@ public class EarningPeriodsValidationProcessorTest
         private List<DataLockFailure> dataLockFailures;
```
[the remark this answers is not in the document]
Comment (abroodie) [reply src/SFA.DAS.Payments.DataLocks.Domain.UnitTests/Services/CourseValidation/EarningPeriodsValidationProcessorTests.cs]: Done

File: src/SFA.DAS.Payments.DataLocks.DataLockService/DataLockService.cs [hunk 21]
```diff
@@ -1,107 +1,148 @@
...
+            Func<IApprenticeshipRepository> apprenticeshipRepository,
+            IActorDataCache<List<ApprenticeshipModel>> apprenticeships,
+            IActorDataCache<List<long>> providers,
+            IDataLockProcessor dataLockProcessor,
+            IApprenticeshipUpdatedProcessor apprenticeshipUpdatedProcessor,
+            ITelemetry telemetry
+            )
+            : base(actorService, actorId)
+        {
+            this.actorService = actorService;
```
[the remark this answers is not in the document]
Comment (abroodie) [reply src/SFA.DAS.Payments.DataLocks.DataLockService/DataLockService.cs]: Done

File: src/SFA.DAS.Payments.Monitoring.Jobs.Client/JobMessageClientFactory.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (davehowlett) [reply src/SFA.DAS.Payments.Monitoring.Jobs.Client/JobMessageClientFactory.cs]: PV2-1362 added to the backlog

**Pull request merged**